### PR TITLE
feat: dynamic datasource loading

### PR DESCRIPTION
**Pull request by zharinov**
Ref #5419

### CODE REVIEW

File: lib/datasource/index.ts [hunk 1]
```diff
@@ -1,30 +1,8 @@
...
 import * as versioning from '../versioning';
 
-import * as cargo from './cargo';
-import * as cdnjs from './cdnjs';
-import * as dart from './dart';
-import * as docker from './docker';
-import * as hex from './hex';
-import * as github from './github';
-import * as gitlab from './gitlab';
-import * as gitTags from './git-tags';
```
Comment (rarkins) [inline lib/datasource/index.ts]: We now have a mismatch between `git-tags` (autoloaded) and `gitTags`, right? As well as gitSubmodules, gradleVersion, rubyVersion and terraformProvivder.

Comment (zharinov) [reply lib/datasource/index.ts]: Should be fixed now

File: lib/datasource/index.ts [hunk 2]
```diff
@@ -36,31 +14,37 @@ import { VERSION_SCHEME_SEMVER } from '../constants/version-schemes';
...
+  for (const datasourceName of datasourceDirs) {
+    let module = null;
+    try {
+      module = require(`./${datasourceName}`); // eslint-disable-line
+    } catch (err) /* istanbul ignore next */ {
+      logger.fatal({ err }, `Can not load datasource "${datasourceName}".`);
+      process.exit(1);
+    }
+
+    if (isValidDatasourceModule(datasourceName, module)) {
```
Comment (rarkins) [inline lib/datasource/index.ts]: We want to abort if not valid

**Review by rarkins — CHANGES_REQUESTED** [review]
Abort if invalid module loaded

**Comment by rarkins** [comment]
@zharinov this could be a good time to replace the process.exit() with throwing errors?

image

**Comment by rarkins** [comment]
Superseded by #5440

**Comment by rarkins** [comment]
This is presumably failing due to cyclic dependencies, but I'm not sure why/where. Cc @ViceIce

**Comment by ViceIce** [comment]
🤔 I'll look at it next, finishing docker caching first

**Comment by zharinov** [comment]
For each manager: Cannot find module './__snapshots__' from 'index.ts'

**Comment by ViceIce** [comment]
ahh, we need to filter folders starting with `__` double underscore

**Pull request merged**